File: objc/source/abstract/operable.cpp
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <WCDB/operable.hpp>
#include <WCDB/literal_value.hpp>
#include <WCDB/expression.hpp>
#include <WCDB/convertible_impl.hpp>
#include <WCDB/utility.hpp>

namespace WCDB {
    
Expression Operator::operateWithRaw(const std::string &raw)
{
    return Expression(raw, nullptr);
}

Expression Operator::operateWithPrefix(const std::string &prefix, const Expression& operand)
{
    return Expression("(" + prefix + operand.getDescription() + ")", nullptr);
}
    
Expression Operator::operateWithTitle(const std::string &title, const std::string& infix, const std::list<const Expression> &list) {
    return Expression(title + "(" + (infix.empty()?"":infix+" ") + stringByJoiningList(list) + ")", nullptr);
}

Expression Operator::operateWithPostfix(const Expression& operand, const std::string &postfix)
{
    return Expression("(" + operand.getDescription() + " " + postfix + ")", nullptr);
}

Expression Operator::operateWithOperator(const Expression& left, const std::string &op, const Expression& right)
{
    return Expression("(" + left.getDescription() + " " + op + " " + right.getDescription() + ")", nullptr);
}

Expression Operator::operateWithComplexOperator(const Expression& one, const std::string &op1, const Expression& two, const std::string& op2, const Expression &three)
{
    return Expression("(" + one.getDescription() + " " + op1 + " " + two.getDescription() + " " + op2 + " " + three.getDescription() + ")", nullptr);
}

Expression Operable::operator !() const
{
    return Operator::operateWithPrefix("NOT ", asExpression());
}
Expression Operable::operator +() const
{
    return Operator::operateWithPrefix("", asExpression());
}
Expression Operable::operator -() const
{
    return Operator::operateWithPrefix("-", asExpression());
}
Expression Operable::operator ~() const
{
    return Operator::operateWithPrefix("~", asExpression());
}
Expression Operable::operator ||(const Expression& operand) const
{
    return Operator::operateWithOperator(asExpression(), "OR", operand);
}
Expression Operable::operator &&(const Expression& operand) const
{
    return Operator::operateWithOperator(asExpression(), "AND", operand);
}
Expression Operable::operator *(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "*", operand);
}
Expression Operable::operator /(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "/", operand);
}
Expression Operable::operator %(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "%", operand);
}
Expression Operable::operator +(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "+", operand);
}
Expression Operable::operator -(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "-", operand);
}
Expression Operable::operator <<(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "<<", operand);
}
Expression Operable::operator >>(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), ">>", operand);
}
Expression Operable::operator &(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "&", operand);
}
Expression Operable::operator |(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "|", operand);
}
Expression Operable::operator <(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "<", operand);
}
Expression Operable::operator <=(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "<=", operand);
}
Expression Operable::operator >(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), ">", operand);
}
Expression Operable::operator >=(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), ">=", operand);
}
Expression Operable::operator ==(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "=", operand);
}
Expression Operable::operator !=(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "!=", operand);
}
Expression Operable::concat(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "||", operand);
}
Expression Operable::substr(const Expression& start, const Expression& length) const {
    return Operator::operateWithRaw("SUBSTR(" + Expression(asExpression()).getDescription() + ", " + start.getDescription() + ", " + length.getDescription() + ")");
}
Expression Operable::like(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "LIKE", operand);
}
Expression Operable::glob(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "GLOB", operand);
}
Expression Operable::match(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "MATCH", operand);
}
Expression Operable::regexp(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "REGEXP", operand);
}
Expression Operable::notLike(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "NOT LIKE", operand);
}
Expression Operable::notGlob(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "NOT GLOB", operand);
}
Expression Operable::notMatch(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "NOT MATCH", operand);
}
Expression Operable::notRegexp(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "NOT REGEXP", operand);
}
Expression Operable::like(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "LIKE", operand, "ESCAPE", escape);
}
Expression Operable::glob(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "GLOB", operand, "ESCAPE", escape);
}
Expression Operable::match(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "MATCH", operand, "ESCAPE", escape);
}
Expression Operable::regexp(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "REGEXP", operand, "ESCAPE", escape);
}
Expression Operable::notLike(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "NOT LIKE", operand, "ESCAPE", escape);
}
Expression Operable::notGlob(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "NOT GLOB", operand, "ESCAPE", escape);
}
Expression Operable::notMatch(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "NOT MATCH", operand, "ESCAPE", escape);
}
Expression Operable::notRegexp(const Expression& operand, const Expression& escape) const {
    return Operator::operateWithComplexOperator(asExpression(), "NOT REGEXP", operand, "ESCAPE", escape);
}
Expression Operable::isNull() const {
    return Operator::operateWithPostfix(asExpression(), "ISNULL");
}
Expression Operable::isNotNull() const {
    return Operator::operateWithPostfix(asExpression(), "NOTNULL");
}
Expression Operable::is(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "IS", operand);
}
Expression Operable::isNot(const Expression& operand) const {
    return Operator::operateWithOperator(asExpression(), "IS NOT", operand);
}
Expression Operable::between(const Expression& begin, const Expression& end) const {
    return Operator::operateWithComplexOperator(asExpression(), "BETWEEN", begin, "AND", end);
}
Expression Operable::notBetween(const Expression& begin, const Expression& end) const {
    return Operator::operateWithComplexOperator(asExpression(), "NOT BETWEEN", begin, "AND", end);
}
Expression Operable::in(const StatementSelect& statementSelect) const {
    return Operator::operateWithPrefix("IN ", statementSelect);
}
Expression Operable::notIn(const StatementSelect&  statementSelect) const {
    return Operator::operateWithPrefix("NOT IN ", statementSelect);
}
    
Expression Operable::in(const std::list<const Expression> &list) const
{
    return Operator::operateWithPostfix(asExpression(), "IN(" + stringByJoiningList(list) + ")");
}
    
Expression Operable::notIn(const std::list<const Expression> &list) const
{
    return Operator::operateWithPostfix(asExpression(), "NOT IN(" + stringByJoiningList(list) + ")");
}
    
Expression Operable::in(const Expression &expression) const
{
    return Operator::operateWithPostfix(asExpression(), "IN(" + expression.getDescription() + ")");
}
    
Expression Operable::notIn(const Expression &expression) const
{
    return Operator::operateWithPostfix(asExpression(), "NOT IN(" + expression.getDescription() + ")");
}

Expression Operable::avg(bool isDistinct) const {
    return Expression::Function("AVG", asExpression(), isDistinct);;
}
Expression Operable::count(bool isDistinct) const {
    return Expression::Function("COUNT", asExpression(), isDistinct);
}
Expression Operable::groupConcat(bool isDistinct) const {
    return Expression::Function("GROUP_CONCAT", asExpression(), isDistinct);
}
Expression Operable::groupConcat(bool isDistinct, const std::string& seperator) const {
    return Expression::Function("GROUP_CONCAT", {asExpression(), seperator}, isDistinct);
}
Expression Operable::max(bool isDistinct) const {
    return Expression::Function("MAX", asExpression(), isDistinct);
}
Expression Operable::min(bool isDistinct) const {
    return Expression::Function("MIN", asExpression(), isDistinct);
}
Expression Operable::sum(bool isDistinct) const {
    return Expression::Function("SUM", asExpression(), isDistinct);
}
Expression Operable::total(bool isDistinct) const {
    return Expression::Function("TOTAL", asExpression(), isDistinct);
}
Expression Operable::abs(bool isDistinct) const {
    return Expression::Function("ABS", asExpression(), isDistinct);
}
Expression Operable::hex(bool isDistinct) const {
    return Expression::Function("HEX", asExpression(), isDistinct);
}
Expression Operable::length(bool isDistinct) const {
    return Expression::Function("LENGTH", asExpression(), isDistinct);
}
Expression Operable::lower(bool isDistinct) const {
    return Expression::Function("LOWER", asExpression(), isDistinct);
}
Expression Operable::upper(bool isDistinct) const {
    return Expression::Function("UPPER", asExpression(), isDistinct);
}
Expression Operable::round(bool isDistinct) const {
    return Expression::Function("ROUND", asExpression(), isDistinct);
}
Expression Operable::matchinfo() const {
    return Expression::Function("MATCHINFO", asExpression());
}
Expression Operable::offsets() const {
    return Expression::Function("OFFSETS", asExpression());
}
Expression Operable::snippet() const {
    return Expression::Function("SNIPPET", asExpression());
}
    
} //namespace WCDB

File: swift/source/abstract/Operable.swift
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

extension ExpressionConvertible {
    static func operate(prefix: String, operand: ExpressionConvertible) -> Expression {
        return Expression(withRaw: "(\(prefix)\(operand.asExpression().description))")
    }
    static func operate(title: String, infix: String?, operands: [ExpressionConvertible]) -> Expression {
        return Expression(withRaw: "\(title)(\(infix != nil ? infix!+" " : "")\(operands.joined()))")
    }
    static func operate(operand: ExpressionConvertible, postfix: String) -> Expression {
        return Expression(withRaw: "(\(operand.asExpression().description) \(postfix))")
    }
    static func operate(left: ExpressionConvertible,
                        `operator`: String,
                        right: ExpressionConvertible) -> Expression {
        let leftDescription = left.asExpression().description
        let rightDescription = right.asExpression().description
        return Expression(withRaw: "(\(leftDescription) \(`operator`) \(rightDescription))")
    }
    static func operate(one: ExpressionConvertible,
                        operator1: String,
                        two: ExpressionConvertible,
                        operator2: String,
                        three: ExpressionConvertible) -> Expression {
        let oneDescription = one.asExpression().description
        let twoDescription = two.asExpression().description
        let threeDescription = three.asExpression().description
        let raws = [oneDescription, operator1, twoDescription, operator2, threeDescription]
        return Expression(withRaw: "(\(raws.joined(separateBy: " ")))")
    }
}

public protocol ExpressionCanBeOperated: ExpressionConvertible {
    static func || <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func && <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func * <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func / <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func % <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func + <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func - <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func << <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func >> <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func & <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func | <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func < <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func <= <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func > <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func >= <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func == <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression

    static func != <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression
}

extension ExpressionCanBeOperated {
    public static func || <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return Expression.operate(left: left, operator: "OR", right: right)
    }

    public static func && <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "AND", right: right)
    }

    public static func * <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "*", right: right)
    }

    public static func / <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "/", right: right)
    }

    public static func % <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "%", right: right)
    }

    public static func + <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "+", right: right)
    }

    public static func - <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "-", right: right)
    }

    public static func << <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "<<", right: right)
    }

    public static func >> <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: ">>", right: right)
    }

    public static func & <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "&", right: right)
    }

    public static func | <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "|", right: right)
    }

    public static func < <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "<", right: right)
    }

    public static func <= <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "<=", right: right)
    }

    public static func > <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: ">", right: right)
    }

    public static func >= <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: ">=", right: right)
    }

    public static func == <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "=", right: right)
    }

    public static func != <ExpressionOperableType: ExpressionOperable>(
        left: Self, right: ExpressionOperableType) -> Expression {
        return operate(left: left, operator: "!=", right: right)
    }
}

public protocol ExpressionOperable: ExpressionCanBeOperated {
    //Unary
    prefix static func ! (operand: Self) -> Expression
    prefix static func + (operand: Self) -> Expression
    prefix static func - (operand: Self) -> Expression
    prefix static func ~ (operand: Self) -> Expression

    //Binary
    static func || <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self,
        right: ExpressionConvertibleType) -> Expression

    static func && <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func * <ExpressionConvertibleType: ExpressionConvertible>
        (left: Self, right: ExpressionConvertibleType) -> Expression

    static func / <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func % <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func + <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func - <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func << <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func >> <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func & <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func | <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func < <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func <= <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func > <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func >= <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func == <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    static func != <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression

    func concat(_ operand: ExpressionConvertible) -> Expression
    func substr(start: ExpressionConvertible, length: ExpressionConvertible) -> Expression

    func like(_ operand: ExpressionConvertible) -> Expression
    func glob(_ operand: ExpressionConvertible) -> Expression
    func match(_ operand: ExpressionConvertible) -> Expression
    func regexp(_ operand: ExpressionConvertible) -> Expression

    func notLike(_ operand: ExpressionConvertible) -> Expression
    func notGlob(_ operand: ExpressionConvertible) -> Expression
    func notMatch(_ operand: ExpressionConvertible) -> Expression
    func notRegexp(_ operand: ExpressionConvertible) -> Expression

    func like(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression
    func glob(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression
    func match(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression
    func regexp(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression

    func notLike(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression
    func notGlob(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression
    func notMatch(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression
    func notRegexp(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression

    func isNull() -> Expression
    func isNotNull() -> Expression

    func `is`(_ operand: ExpressionConvertible) -> Expression
    func isNot(_ operand: ExpressionConvertible) -> Expression

    func between(_ begin: ExpressionConvertible, _ end: ExpressionConvertible) -> Expression
    func notBetween(_ begin: ExpressionConvertible, _ end: ExpressionConvertible) -> Expression

    func `in`(_ statementSelect: StatementSelect) -> Expression
    func notIn(_ statementSelect: StatementSelect) -> Expression

    func `in`(_ expressionConvertibleList: ExpressionConvertible...) -> Expression
    func notIn(_ expressionConvertibleList: ExpressionConvertible...) -> Expression

    func `in`(_ expressionConvertibleList: [ExpressionConvertible]) -> Expression
    func notIn(_ expressionConvertibleList: [ExpressionConvertible]) -> Expression

    //aggregate functions
    func avg(isDistinct: Bool) -> Expression
    func count(isDistinct: Bool) -> Expression
    func groupConcat(isDistinct: Bool) -> Expression
    func groupConcat(isDistinct: Bool, separateBy seperator: String) -> Expression
    func max(isDistinct: Bool) -> Expression
    func min(isDistinct: Bool) -> Expression
    func sum(isDistinct: Bool) -> Expression
    func total(isDistinct: Bool) -> Expression

    //core functions
    func abs(isDistinct: Bool) -> Expression
    func hex(isDistinct: Bool) -> Expression
    func length(isDistinct: Bool) -> Expression
    func lower(isDistinct: Bool) -> Expression
    func upper(isDistinct: Bool) -> Expression
    func round(isDistinct: Bool) -> Expression

    //FTS3
    func matchinfo() -> Expression
    func offsets() -> Expression
    func snippet() -> Expression
}

extension ExpressionOperable {
    //Unary
    public prefix static func ! (operand: Self) -> Expression {
        return operate(prefix: "NOT ", operand: operand)
    }
    public prefix static func + (operand: Self) -> Expression {
        return operate(prefix: "", operand: operand)
    }
    public prefix static func - (operand: Self) -> Expression {
        return operate(prefix: "-", operand: operand)
    }
    public prefix static func ~ (operand: Self) -> Expression {
        return operate(prefix: "~", operand: operand)
    }

    //Binary
    public static func || <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "OR", right: right)
    }

    public static func && <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "AND", right: right)
    }

    public static func * <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "*", right: right)
    }

    public static func / <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "/", right: right)
    }

    public static func % <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "%", right: right)
    }

    public static func + <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "+", right: right)
    }

    public static func - <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "-", right: right)
    }

    public static func << <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "<<", right: right)
    }

    public static func >> <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: ">>", right: right)
    }

    public static func & <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "&", right: right)
    }

    public static func | <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "|", right: right)
    }

    public static func < <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "<", right: right)
    }

    public static func <= <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "<=", right: right)
    }

    public static func > <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: ">", right: right)
    }

    public static func >= <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: ">=", right: right)
    }

    public static func == <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "=", right: right)
    }

    public static func != <ExpressionConvertibleType: ExpressionConvertible>(
        left: Self, right: ExpressionConvertibleType) -> Expression {
        return operate(left: left, operator: "!=", right: right)
    }

    public func concat(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "||", right: operand)
    }
    public func substr(start: ExpressionConvertible, length: ExpressionConvertible) -> Expression {
        let description = asExpression().description
        let startDescription = start.asExpression().description
        let lengthDescription = length.asExpression().description
        return Expression(withRaw: "SUBSTR(\(description), \(startDescription), \(lengthDescription))")
    }

    public func like(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "LIKE", right: operand)
    }
    public func glob(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "GLOB", right: operand)
    }
    public func match(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "MATCH", right: operand)
    }
    public func regexp(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "REGEXP", right: operand)
    }
    public func notLike(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "NOT LIKE", right: operand)
    }
    public func notGlob(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "NOT GLOB", right: operand)
    }
    public func notMatch(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "NOT MATCH", right: operand)
    }
    public func notRegexp(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "NOT REGEXP", right: operand)
    }

    public func like(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "LIKE", two: operand, operator2: "ESCAPE", three: escape)
    }
    public func glob(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "GLOB", two: operand, operator2: "ESCAPE", three: escape)
    }
    public func match(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "MATCH", two: operand, operator2: "ESCAPE", three: escape)
    }
    public func regexp(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "REGEXP", two: operand, operator2: "ESCAPE", three: escape)
    }
    public func notLike(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "NOT LIKE", two: operand, operator2: "ESCAPE", three: escape)
    }
    public func notGlob(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "NOT GLOB", two: operand, operator2: "ESCAPE", three: escape)
    }
    public func notMatch(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "NOT MATCH", two: operand, operator2: "ESCAPE", three: escape)
    }
    public func notRegexp(_ operand: ExpressionConvertible, escape: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "NOT REGEXP", two: operand, operator2: "ESCAPE", three: escape)
    }

    public func isNull() -> Expression {
        return Self.operate(operand: self, postfix: "ISNULL")
    }
    public func isNotNull() -> Expression {
        return Self.operate(operand: self, postfix: "NOTNULL")
    }
    public func `is`(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "IS", right: operand)
    }
    public func isNot(_ operand: ExpressionConvertible) -> Expression {
        return Self.operate(left: self, operator: "IS NOT", right: operand)
    }
    public func between(_ begin: ExpressionConvertible, _ end: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "BETWEEN", two: begin, operator2: "AND", three: end)
    }
    public func notBetween(_ begin: ExpressionConvertible, _ end: ExpressionConvertible) -> Expression {
        return Self.operate(one: self, operator1: "NOT BETWEEN", two: begin, operator2: "AND", three: end)
    }

    public func `in`(_ statementSelect: StatementSelect) -> Expression {
        return Self.operate(prefix: "IN ", operand: statementSelect)
    }
    public func notIn(_ statementSelect: StatementSelect) -> Expression {
        return Self.operate(prefix: "NOT IN ", operand: statementSelect)
    }
    public func `in`(_ expressionConvertibleList: ExpressionConvertible...) -> Expression {
        return self.`in`(expressionConvertibleList)
    }
    public func notIn(_ expressionConvertibleList: ExpressionConvertible...) -> Expression {
        return self.notIn(expressionConvertibleList)
    }
    public func `in`(_ expressionConvertibleList: [ExpressionConvertible]) -> Expression {
        return Self.operate(operand: self, postfix: "IN(\(expressionConvertibleList.joined()))")
    }
    public func notIn(_ expressionConvertibleList: [ExpressionConvertible]) -> Expression {
        return Self.operate(operand: self, postfix: "NOT IN(\(expressionConvertibleList.joined()))")
    }

    //aggregate functions
    public func avg(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "AVG", self, isDistinct: isDistinct)
    }
    public func count(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "COUNT", self, isDistinct: isDistinct)
    }
    public func groupConcat(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "GROUP_CONCAT", self, isDistinct: isDistinct)
    }
    public func groupConcat(isDistinct: Bool = false, separateBy seperator: String) -> Expression {
        return Expression.function(named: "GROUP_CONCAT", self, seperator, isDistinct: isDistinct)
    }
    public func max(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "MAX", self, isDistinct: isDistinct)
    }
    public func min(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "MIN", self, isDistinct: isDistinct)
    }
    public func sum(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "SUM", self, isDistinct: isDistinct)
    }
    public func total(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "TOTAL", self, isDistinct: isDistinct)
    }
    //core functions
    public func abs(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "ABS", self, isDistinct: isDistinct)
    }
    public func hex(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "HEX", self, isDistinct: isDistinct)
    }
    public func length(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "LENGTH", self, isDistinct: isDistinct)
    }
    public func lower(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "LOWER", self, isDistinct: isDistinct)
    }
    public func upper(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "UPPER", self, isDistinct: isDistinct)
    }
    public func round(isDistinct: Bool = false) -> Expression {
        return Expression.function(named: "ROUND", self, isDistinct: isDistinct)
    }

    //FTS3
    public func matchinfo() -> Expression {
        return Expression.function(named: "MATCHINFO", self)
    }
    public func offsets() -> Expression {
        return Expression.function(named: "OFFSETS", self)
    }
    public func snippet() -> Expression {
        return Expression.function(named: "SNIPPET", self)
    }
}
